### ISSUE TITLE
decode arguments into enums

### ISSUE THREAD

**Issue by rukai** [tailhook/knuffel#19]
I have a KDL document like:
```kdl
pad {
    left "Left"
    right "Right"
    up "Up"
    down "Down"
}
```

And I want to decode it into the following structure.
```rust
#[derive(knuffel::Decode)]
pub struct Config {
    #[knuffel(child)]
    pub pad: Pad,
}

#[derive(knuffel::Decode)]
pub struct Pad {
    #[knuffel(child, str, unwrap(argument))]
    pub left: KeyMap,
    #[knuffel(child, str, unwrap(argument))]
    pub right: KeyMap,
    #[knuffel(child, str, unwrap(argument))]
    pub up: KeyMap,
    #[knuffel(child, str, unwrap(argument))]
    pub down: KeyMap,
}

#[derive(enum_utils::FromStr)]
pub enum KeyMap {
    Left,
    Right,
    Up,
    Down
}
```

This is a very common pattern when working with serde but it seems knuffel doesnt support enums as arguments/properties very well?
I tried the knuffel derive attributes that I included in this example as well as a few others but all of them either failed to compile or failed to parse the kdl.

**Comment by rnarkk** [comment]
**TL;DR**: Jump to 3.

Before anything else, I would have to tell you that your design is wrong. And this is being more or less caused by knuffel’s own misconception about children #18 and modelling methodologies that push you to resort to the `unwrap` behaviour. To explain one by one, in KDL,

1. First of all, you don’t have to include arguments or properties. If only those existences have meaning, your nodes can be without them:

```kdl
pad {
    left
    right
    up
    down
}
```

2. Second, and this is where knuffel doesn’t correctly conceive the idea, child nodes are basically for enumeration; in your case, you would want either

2.1. to allow duplicate `Key`s (now better to be called this way instead of `KeyMap`s) in `Pad`:

```kdl
pad {
  left
  left
  up
}
```

```rust
#[derive(Decode)]
struct Pad {
    #[knuffel(children)]
    keys: Vec<Key>
}

#[derive(Decode)]
enum Key {
    ..
}
```

2.2. or to wind up `Key` to the level of `Config`:

```kdl
left
```

```rust
#[derive(Decode)]
struct Config {
    #[knuffel(child)]
    key: Key
}
```

But currently this doesn’t work due to knuffel’s wrong implementation by the misconception.

3. Lastly, and this is most likely what you initially wanted, you can turn enum variants into string values by implementing `DecodeScalar`:

```kdl
key "left"
```

```rust
#[derive(Decode)]
struct Pad {
    #[knuffel(argument)]
    key: Key
}

#[derive(DecodeScalar)]
enum Key {
    ..
}
```

**Comment by tailhook** [comment]
Hi @rukai.

Sorry for the long delay, it works you just have to use `str` in `unwrap`:
```
    #[knuffel(child, str, unwrap(argument))]
    pub left: KeyMap,
```
This is because `str` works on the argument that is going to be unwrapped. I.e. we first parse thing like `#[knuffel(argument, str)]` and then do `unwrap`.

The alternative is also use knuffel's derive
```
#[derive(knuffel::DecodeScalar)]
pub enum KeyMap {
    Left,
```
This gets better error reporting in my opinion, but we don't have `rename` option for that yet.

I'll take a look why `child, str` doesn't fail. But other than that there is nothing to do here. Feel free to open PR for documentation or another feature request on improving DecodeScalar.

**Comment by tailhook** [comment]
Error message improved in #25:
```
